File: shared/src/androidMain/kotlin/com/snabbit/runner/shared/core/camera/CapturePathHandler.kt
package com.snabbit.runner.shared.core.camera

import android.webkit.WebResourceResponse
import androidx.webkit.WebViewAssetLoader
import com.snabbit.runner.shared.core.Logger
import java.io.File
import java.io.FileInputStream
import java.util.concurrent.ConcurrentHashMap

/**
 * Serves captured images and videos from disk via `WebViewAssetLoader` so that
 * a URL like `https://appassets.androidplatform.net/captures/<token>.jpg`
 * is intercepted by the WebView's native networking layer and streamed
 * directly from the app-private file — **no media bytes cross the JS bridge**.
 *
 * Implements [CaptureIndex]: construct the [CaptureRegistry] with this handler as
 * its `index`, so the registry drives put/remove and this class holds the ONE
 * synchronous token→path map the WebView callback reads. (The registry's own
 * [CaptureRegistry.resolve] is `suspend` and can't be called from the callback.)
 *
 * Ported from Flutter's `capture_path_handler.dart`. Registered at `/captures/`.
 *
 * @param capturesDir when non-null, only files resolving *under* this directory
 *   are served — defense in depth for the day a token is ever derived into a path
 *   (today it's an opaque map key, so traversal is already impossible).
 */
class CapturePathHandler(
    private val logger: Logger,
    private val capturesDir: File? = null,
) : WebViewAssetLoader.PathHandler, CaptureIndex {

    /**
     * Synchronous token→path mirror. `WebViewAssetLoader.PathHandler.handle()`
     * runs on a WebView **worker thread** (not the main thread) while registry
     * updates arrive from the coroutine world — so this MUST be thread-safe.
     */
    private val registeredPaths = ConcurrentHashMap<String, String>()

    override fun put(token: String, filePath: String) {
        registeredPaths[token] = filePath
    }

    override fun remove(token: String) {
        registeredPaths.remove(token)
    }

    override fun handle(path: String): WebResourceResponse? {
        // Path arrives as e.g. "1716825600000_0.jpg" or "1716825600000_0.mp4".
        val filename = path.substringAfterLast("/")
        val token: String
        val contentType: String
        val maxSize: Long

        when {
            filename.endsWith(".jpg") || filename.endsWith(".jpeg") -> {
                token = filename.substringBeforeLast(".")
                contentType = "image/jpeg"
                maxSize = CaptureRegistry.MAX_PHOTO_SIZE_BYTES
            }
            filename.endsWith(".mp4") -> {
                token = filename.substringBeforeLast(".")
                contentType = "video/mp4"
                maxSize = CaptureRegistry.MAX_VIDEO_SIZE_BYTES
            }
            else -> {
                // No extension — assume photo token.
                token = filename
                contentType = "image/jpeg"
                maxSize = CaptureRegistry.MAX_PHOTO_SIZE_BYTES
            }
        }

        val filePath = registeredPaths[token]
        if (filePath == null) {
            logger.d(TAG, "Serve miss: token=$token, reason=unknown_or_expired")
            return notFound()
        }

        val file = File(filePath)
        // Must be a regular file, and (when a root is configured) resolve under it.
        if (!file.isFile || !file.isWithin(capturesDir)) {
            logger.d(TAG, "Serve miss: token=$token, reason=file_missing_or_out_of_scope")
            return notFound()
        }

        val fileLength = file.length()
        if (fileLength > maxSize) {
            logger.w(TAG, "Serve rejected: token=$token, size=$fileLength, max=$maxSize")
            return notFound()
        }

        logger.d(TAG, "Serve hit: token=$token, size=$fileLength, type=$contentType")

        // Served same-origin inside the app's own WebView — no CORS headers, and
        // the token is never echoed back.
        return WebResourceResponse(
            contentType,
            null, // encoding — null for binary
            200,
            "OK",
            mapOf(
                "Cache-Control" to "no-store",
                "Content-Length" to fileLength.toString(),
            ),
            FileInputStream(file),
        )
    }

    /** True if [root] is null, or this file canonically resolves under [root]. */
    private fun File.isWithin(root: File?): Boolean {
        if (root == null) return true
        return runCatching {
            canonicalPath.startsWith(root.canonicalPath + File.separator)
        }.getOrDefault(false)
    }

    companion object {
        private const val TAG = "CapturePathHandler"

        /** 404 with no body — no CORS headers (same-origin only). */
        private fun notFound(): WebResourceResponse =
            WebResourceResponse("text/plain", "utf-8", 404, "Not Found", emptyMap(), null)
    }
}

File: lib/services/webview/capture_path_handler.dart
import 'package:flutter_inappwebview/flutter_inappwebview.dart';
import 'package:snabbit_runner/services/monitoring/monitoring_service_helper.dart';
import 'package:snabbit_runner/services/webview/capture_registry.dart';

/// A [CustomPathHandler] that serves captured images from the
/// [CaptureRegistry] via `WebViewAssetLoader`.
///
/// Registered at `/captures/` so a URL like
/// `https://appassets.androidplatform.net/captures/<token>.jpg`
/// is intercepted by the WebView's native networking layer and served
/// directly from disk — **no image bytes cross the JS bridge**.
///
/// Returns a fully-specified [WebResourceResponse] (status code, headers,
/// Content-Length) so the WebView's image decoder can stream efficiently.
/// Unknown or expired tokens receive a 404 (null data).
///
/// **CORS note:** All responses include `Access-Control-Allow-Origin: *`
/// and permissive `Access-Control-Allow-Methods/Headers` so the web's
/// `fetch()` works cross-origin. [CustomPathHandler.handle] only receives
/// the path — the HTTP method is not exposed, so we cannot return a
/// dedicated 204 for OPTIONS preflight. The headers on every response
/// satisfy preflight requirements on Android WebView. If the web team
/// adds custom request headers, verify preflight still works on the
/// target WebView versions.
class CapturePathHandler extends CustomPathHandler {
  CapturePathHandler({required super.path});

  @override
  Future<WebResourceResponse?> handle(String path) async {
    // Path arrives as e.g. "1716825600000000_0.jpg" — strip the trailing
    // extension only (not greedy like replaceAll).
    final filename = path.split('/').last;
    final token = filename.endsWith('.jpg')
        ? filename.substring(0, filename.length - 4)
        : filename;
    final stopwatch = Stopwatch()..start();

    final file = CaptureRegistry.instance.resolve(token);

    if (file == null || !await file.exists()) {
      stopwatch.stop();
      MonitoringServiceHelper.logInfo(
        'capture_serve_miss',
        {
          'token': token,
          'reason': file == null ? 'unknown_or_expired' : 'file_deleted',
          'latencyMs': stopwatch.elapsedMilliseconds,
          'timestamp': DateTime.now().toIso8601String(),
        },
      );
      return _notFound(token);
    }

    // Read + serve, wrapped to close a TOCTOU window: the registry's sweep
    // timer or a web `releaseCapture` can delete the file between the
    // resolve()/exists() check above and the read below. Without this guard
    // the exception propagates into the native WebView layer untracked;
    // here we treat it as a logged miss instead.
    try {
      // Size guard before loading into memory — WebResourceResponse only
      // accepts Uint8List (no streaming API on Android). Capture-time
      // enforcement in CaptureImageHandler should mean we never reach this,
      // so it's defense-in-depth; we return a distinct 413 (not 404) so the
      // cause is observable and the web could differentiate if needed.
      final fileLength = await file.length();
      if (fileLength > CaptureRegistry.maxCaptureSizeBytes) {
        stopwatch.stop();
        MonitoringServiceHelper.logWarning(
          'capture_serve_too_large',
          {
            'token': token,
            'fileSizeBytes': fileLength,
            'maxAllowed': CaptureRegistry.maxCaptureSizeBytes,
            'timestamp': DateTime.now().toIso8601String(),
          },
        );
        return _tooLarge(token);
      }

      final bytes = await file.readAsBytes();
      stopwatch.stop();

      MonitoringServiceHelper.logDebug(
        'capture_serve_hit',
        {
          'token': token,
          'fileSizeBytes': bytes.length,
          'latencyMs': stopwatch.elapsedMilliseconds,
          'timestamp': DateTime.now().toIso8601String(),
        },
      );

      return WebResourceResponse(
        contentType: 'image/jpeg',
        statusCode: 200,
        reasonPhrase: 'OK',
        headers: {
          ..._corsHeaders,
          'Cache-Control': 'no-store',
          'Content-Length': '${bytes.length}',
          'X-Capture-Token': token,
        },
        data: bytes,
      );
    } catch (e) {
      stopwatch.stop();
      // A thrown read is an error condition (I/O failure, or the TOCTOU
      // delete race) — distinct from a clean not-found, so log at error
      // level under its own event rather than as a benign miss.
      MonitoringServiceHelper.logError(
        'capture_serve_error',
        {
          'token': token,
          'reason': 'read_failed',
          'error': e.toString(),
          'latencyMs': stopwatch.elapsedMilliseconds,
          'timestamp': DateTime.now().toIso8601String(),
        },
      );
      return _notFound(token);
    }
  }

  /// CORS headers included on every response. Permissive enough to
  /// satisfy both simple requests and preflight checks on Android
  /// WebView (where we cannot intercept OPTIONS separately).
  ///
  /// **API-level note:** The wildcard `*` in `Allow-Headers` is
  /// supported from Chromium 87+ (Android API 30+). Devices on
  /// API 28–29 with older system WebView builds may silently ignore
  /// the wildcard. This is acceptable here because the asset-loader
  /// domain is same-origin (`appassets.androidplatform.net`) and the
  /// web's `fetch()` for captures uses no custom request headers.
  /// If custom headers are added later, enumerate them explicitly.
  static const Map<String, String> _corsHeaders = {
    'Access-Control-Allow-Origin': '*',
    'Access-Control-Allow-Methods': 'GET, OPTIONS',
    'Access-Control-Allow-Headers': '*',
  };

  /// Returns a 404 response with CORS headers.
  static WebResourceResponse _notFound(String token) {
    return WebResourceResponse(
      statusCode: 404,
      reasonPhrase: 'Not Found',
      headers: _corsHeaders,
      data: null,
    );
  }

  /// Returns a 413 (Payload Too Large) response — distinct from the 404 a
  /// missing/expired token returns, so the web can tell "image too large,
  /// retake" apart from "token gone".
  static WebResourceResponse _tooLarge(String token) {
    return WebResourceResponse(
      statusCode: 413,
      reasonPhrase: 'Payload Too Large',
      headers: _corsHeaders,
      data: null,
    );
  }
}
